File: frontend/src/model/WasteOrder.ts
import { Service, ServiceSchema } from "./Service";
import { TaskSite, TaskSiteSchema } from "./TaskSite";
import * as Joi from '@hapi/joi';

export enum WasteOrderStatus {
    COMMISSIONED,
    ACCEPTED,
    REJECTED,
    CANCELLED,
    COMPLETED
}

export interface WasteOrder {
    key: string;
    status: WasteOrderStatus;
    subcontractorMSPID: string;
    originatorMSPID: string;
    customerName: string;
    taskSite: TaskSite;
    service: Service;
    description: string;
    quantity: number;
    unitPrice: number;
    unitOfMeasure: string;
    taskDate: Date
    startingTime: string;
    finishingTime: string;
    referenceNo: string;
    rejectionMessage: string;
    lastChanged: Date;
    lastChangedByMSPID: string;
}

export const WasteOrderSchema = Joi.object().keys({
    key: ServiceSchema.required(),
    status: Joi.number().required(),
    subcontractorMSPID: Joi.string().required(),
    originatorMSPID: Joi.string().required(),
    customerName: Joi.string().required(),
    taskSite: TaskSiteSchema.required(),
    service: ServiceSchema.required(),
    description: Joi.string().required(),
    quantity: Joi.number().required(),
    unitPrice: Joi.number().required(),
    unitOfMeasure: Joi.string().required(),
    taskDate: Joi.date().required(),
    startingTime: Joi.string(),
    finishingTime: Joi.string(),
    referenceNo: Joi.string().required(),
    rejectionMessage: Joi.string(),
    lastChanged: Joi.date().required(),
    lastChangedByMSPID: Joi.string().required()
});

export const WasteOrderCommissionSchema = Joi.object().keys({
    subcontractorMSPID: Joi.string().required(),
    customerName: Joi.string().required(),
    taskSite: TaskSiteSchema.required(),
    service: ServiceSchema.required(),
    description: Joi.string().required(),
    quantity: Joi.number().required(),
    unitPrice: Joi.number().required(),
    unitOfMeasure: Joi.string().required(),
    taskDate: Joi.date().required(),
    startingTime: Joi.string(),
    finishingTime: Joi.string(),
    referenceNo: Joi.string().required()
});

export const WasteOrderRecommissionSchema = Joi.object().keys({
    subcontractorMSPID: Joi.string().required(),
    customerName: Joi.string().required(),
    taskSite: TaskSiteSchema.required(),
    service: ServiceSchema.required(),
    description: Joi.string().required(),
    quantity: Joi.number().required(),
    unitPrice: Joi.number().required(),
    unitOfMeasure: Joi.string().required(),
    taskDate: Joi.date().required(),
    startingTime: Joi.string(),
    finishingTime: Joi.string()
});

export const WasteOrderCompleteSchema = Joi.object().keys({
    service: ServiceSchema.required(),
    description: Joi.string().required(),
    quantity: Joi.number().required(),
    unitPrice: Joi.number().required(),
    unitOfMeasure: Joi.string().required(),
    taskDate: Joi.date().required(),
    startingTime: Joi.string(),
    finishingTime: Joi.string()
});

export const WasteOrderRejectSchema = Joi.object().keys({
    rejectionMessage: Joi.string().required()
});

File: mobile/lib/wasteOrder/WasteOrder.dart
import 'package:wastechain_mobile/wasteOrder/Service.dart';
import 'package:wastechain_mobile/wasteOrder/TaskSite.dart';

class WasteOrder {
  final String id;
  final int status;
  final String subcontractorMSPID;
  final String originatorMSPID;
  final String customerName;
  final TaskSite taskSite;
  final Service service;
  final String description;
  final double quantity;
  final double unitPrice;
  final String unitOfMeasure;
  final String taskDate;
  final String startingTime;
  final String finishingTime;
  final String referenceNo;
  final String rejectionMessage;
  final String lastChanged;
  final String lastChangedByMSPID;

  WasteOrder(
      {this.id,
      this.status,
      this.subcontractorMSPID,
      this.originatorMSPID,
      this.customerName,
      this.taskSite,
      this.service,
      this.description,
      this.quantity,
      this.unitPrice,
      this.unitOfMeasure,
      this.taskDate,
      this.startingTime,
      this.finishingTime,
      this.referenceNo,
      this.rejectionMessage,
      this.lastChanged,
      this.lastChangedByMSPID});

  factory WasteOrder.fromJSON(Map<String, dynamic> json) {
    return WasteOrder(
      id: json['id'],
      status: json['status'],
      subcontractorMSPID: json['subcontractorMSPID'],
      originatorMSPID: json['originatorMSPID'],
      customerName: json['customerName'],
      taskSite: TaskSite.fromJSON(json['taskSite']),
      service: Service.fromJSON(json['service']),
      description: json['description'],
      quantity: json['quantity'].toDouble(),
      unitPrice: json['unitPrice'].toDouble(),
      unitOfMeasure: json['unitOfMeasure'],
      taskDate: json['taskDate'],
      startingTime: json['startingTime'],
      finishingTime: json['finishingTime'],
      referenceNo: json['referenceNo'],
      rejectionMessage: json['rejectionMessage'],
      lastChanged: json['lastChanged'],
      lastChangedByMSPID: json['lastChangedByMSPID'],
    );
  }
}
